contracts/andromeda_vault/src/contract.rs

```diff
@@ -0,0 +1,426 @@
+    let withdraw_exec = to_binary(&ExecuteMsg::AndrReceive(AndromedaMsg::Withdraw {
+        recipient: Some(recipient),
+        tokens_to_withdraw: Some(withdrawals),
+    }))?;
+    let withdraw_submsg = SubMsg {
+        id: 104,
+        msg: CosmosMsg::Wasm(WasmMsg::Execute {
+            contract_addr: addr,
+            msg: withdraw_exec,
+            funds: vec![],
+        }),
+        gas_limit: None,
+        reply_on: ReplyOn::Error,
+    };
+
+    Ok(res.add_submessage(withdraw_submsg))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn migrate(deps: DepsMut, _env: Env, _msg: MigrateMsg) -> Result<Response, ContractError> {
+    let version = get_contract_version(deps.storage)?;
+    if version.contract != CONTRACT_NAME {
+        return Err(ContractError::CannotMigrate {
+            previous_contract: version.contract,
+        });
+    }
+    Ok(Response::default())
+}
+
+#[entry_point]
+pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> Result<Binary, ContractError> {
+    match msg {
+        QueryMsg::AndrQuery(msg) => handle_andromeda_query(deps, env, msg),
+        QueryMsg::Balance {
+            address,
+            strategy,
+            denom,
+        } => query_balance(deps, env, address, strategy, denom),
+        QueryMsg::StrategyAddress { strategy } => query_strategy_address(deps, env, strategy),
+    }
+}
+
+fn handle_andromeda_query(
+    deps: Deps,
+    env: Env,
+    msg: AndromedaQuery,
+) -> Result<Binary, ContractError> {
+    ADOContract::default().query(deps, env, msg, query)
+}
+
+fn query_balance(
+    deps: Deps,
+    _env: Env,
+    address: String,
+    strategy: Option<StrategyType>,
+    denom: Option<String>,
+) -> Result<Binary, ContractError> {
+    if let Some(strategy) = strategy {
+        let strategy_addr = STRATEGY_CONTRACT_ADDRESSES.load(deps.storage, strategy.to_string())?;
+        // DEV NOTE: Why does this require a generic type when not using custom query?
+        let query: QueryRequest<Empty> = QueryRequest::Wasm(WasmQuery::Smart {
+            contract_addr: strategy_addr,
+            msg: to_binary(&AndrQueryMsg::AndrQuery(AndromedaQuery::Get(Some(
+                to_binary(&address)?,
+            ))))?,
+        });
+        match deps.querier.raw_query(&to_binary(&query)?.to_vec()) {
+            SystemResult::Ok(ContractResult::Ok(value)) => Ok(value),
+            _ => Err(ContractError::InvalidQuery {}),
+        }
+    } else if let Some(denom) = denom {
+        let balance = BALANCES.load(deps.storage, (address, denom.clone()))?;
+        Ok(to_binary(&[Coin {
+            denom,
+            amount: balance,
+        }])?)
+    } else {
+        let balances: Vec<Coin> = BALANCES
+            .prefix(address)
+            .range(deps.storage, None, None, Order::Ascending)
+            .map(|v| {
+                let (denom_vec, balance) = v.unwrap();
+                let denom = String::from_utf8(denom_vec).unwrap();
```

Comment:
As much as it makes the code more messy, I think we should not unwrap here. If we ever have this fail on us we will be stuck for ages.

Reply:
What's the alternative?

Reply:
This seems to work, not as messy as I thought actually :)
```
        let balances: Result<Vec<Coin>, ContractError> = BALANCES
            .prefix(address)
            .range(deps.storage, None, None, Order::Ascending)
            .flatten()
            .map(|(denom_vec, balance)| {
                let denom = String::from_utf8(denom_vec)?;
                Ok(Coin {
                    denom,
                    amount: balance,
                })
            })
            .collect();
        Ok(to_binary(&balances?)?)
```

Reply:
What this is doing is bubbling up the `Result` which we handle at the end in the `to_binary` call.

Reply:
This might be better as it checks the possible error on `v` and bubbles it up accordingly. This will only error on deserialization, which shouldn't really happen, but good to be safe. 
```
        let balances: Result<Vec<Coin>, ContractError> = BALANCES
            .prefix(address)
            .range(deps.storage, None, None, Order::Ascending)
            .map(|v| {
                let (denom_vec, balance) = v?;
                let denom = String::from_utf8(denom_vec)?;
                Ok(Coin {
                    denom,
                    amount: balance,
                })
            })
            .collect();
        Ok(to_binary(&balances?)?)

```